ninja: build stopped: 'some/path.o' not mentioned in its dyndep file 'some/path/depscan.dd'

**Describe the bug**
The follow error messages appears in my project.
```
ninja: build stopped: 'libscotchDecomp.so.p/plusplus.C.o' not mentioned in its dyndep file 'libscotchDecomp.so.p/depscan.dd'
```
Weirdly, it dissappears if I replace `project('SCOTCH', ['c', 'fortran'])` with `project('SCOTCH', ['c'])`, even though Fortran is never used in the example.

It would be nice if you could either fix the bug, or point me in the right direction to fix it.

Interestingly, this bug only appears if the C++ file ends with .C, not with .cpp. Feel free to say "told you so", as I'm the guy who wrote https://github.com/mesonbuild/meson/pull/8747.

**To Reproduce**
[withFortran.zip](https://github.com/mesonbuild/meson/files/9184433/withFortran.zip)


```
$ cd withFortran
$ meson setup build
$ ninja -C build/ libscotchDecomp.so.p/plusplus.C.o
ninja: Entering directory `build/'
[1/2] Module scanner.
ninja: build stopped: 'libscotchDecomp.so.p/plusplus.C.o' not mentioned in its dyndep file 'libscotchDecomp.so.p/depscan.dd'.
```

**Expected behavior**
`ninja -C build/ libscotchDecomp.so.p/plusplus.C.o` should run without errors, `ninja -C build` should error out.

**system parameters**
* Native Build
* Arch Linux
* Python 3.10.5
* meson 0.63.0
* ninja 1.11.0

--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -975,7 +975,9 @@ class NinjaBackend(backends.Backend):
         all_suffixes = set(compilers.lang_suffixes['cpp']) | set(compilers.lang_suffixes['fortran'])
         selected_sources = []
         for source in compiled_sources:
-            ext = os.path.splitext(source)[1][1:].lower()
+            ext = os.path.splitext(source)[1][1:]
+            if ext != 'C':
+                ext = ext.lower()
             if ext in all_suffixes:
                 selected_sources.append(source)
         return selected_sources
@@ -2706,7 +2708,9 @@ https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47485'''))
         if not self.should_use_dyndeps_for_target(target):
             return
         extension = os.path.splitext(src.fname)[1][1:]
-        if not (extension.lower() in compilers.lang_suffixes['fortran'] or extension in compilers.lang_suffixes['cpp']):
+        if extension != 'C':
+            extension = extension.lower()
+        if not (extension in compilers.lang_suffixes['fortran'] or extension in compilers.lang_suffixes['cpp']):
             return
         dep_scan_file = self.get_dep_scan_file_for(target)
         element.add_item('dyndep', dep_scan_file)

--- a/mesonbuild/scripts/depscan.py
+++ b/mesonbuild/scripts/depscan.py
@@ -51,7 +51,9 @@ class DependencyScanner:
         self.sources_with_exports: T.List[str] = []
 
     def scan_file(self, fname: str) -> None:
-        suffix = os.path.splitext(fname)[1][1:].lower()
+        suffix = os.path.splitext(fname)[1][1:]
+        if suffix != 'C':
+            suffix = suffix.lower()
         if suffix in lang_suffixes['fortran']:
             self.scan_fortran_file(fname)
         elif suffix in lang_suffixes['cpp']:

